beyond-all-reason/Beyond-All-Reason
# Raptor aggro leaderboard

### Work done

This turned into a complete rewrite of the panel file. Sorry! It should be possible to compare the file diff side by side though.
- Most of the non-feature related changes were about reducing cognitive load
- - removing unused variable 
- - reducing if-levels
- - removing unnecessary elses
- - using unreused variables
- - removing logic and strings with no apparent effect
- - reducing negations in conditions and variable names
- Some guidelines taken from here https://springrts.com/wiki/Lua_Performance
- - removing ipairs/pairs
- - localizing
- The leaderboard/playerecoaggro code added
- I believe the resulting code is easier to modify and understand. The parts with the leaderboard drawing and calculation might have added some complexity on this limited part of the file.
- Factored out unitdefs eco calculations so there's no code duplication between gadget and widget (`RaptorCommon` variable)
- Fix text overflow bug on panel first row
- Raptor panel now draws on gamesetup so you can see grace before starting
- Work not done:
- - Removing unused translations
- - Testing/supporting shared id (it should default to only displaying the first player in team)

Flow:
On unitcreated/unitgiven/unitdestroyed the raw eco value is updated by team in a table.
On modulo gameframes this raw data is iterated, expanded with metadata and summed. It is then sorted and iterated on again to created the rendering ready table of objects with string formatting and color/alpha.
On DrawScreen the function iterates the rendering table and truncates strings, handles the font with placement, alignment etc and prints it.

### Setup
Load the branch locally in the `%homepath%\AppData\Local\Programs\Beyond-All-Reason\data\games\BAR.sdd` folder

### Test steps
Starting a game with raptor defense ai is the smallest test. For full test a whole game until queen is killed. The three stages grace, main, queen all affects the results of this widget. Testing different player/AI counts (1-6) is also needed to make sure the layout handles cutting off different amount of players and player names, tinting etc. Eco values are updated on construction start.

Features:
- Aggro is calculated as both the players share/pull with respect to other players (the floating point value) and as a percentage of the whole playerlist (the value inside parantheses)
- If the player or spectated player is below 0.8X eco attraction they will be greyed out
- If the player or spectated player is outside of the top their entry will be greyed out even more
- Red tint on player entry will increase for large aggro values starting at 1.2X maxing out at 6X
- Panel and grace is shown before start
- Top row text does not overflow the panel border
- Grace period timer and queen arrival timer is aligned to the right

before:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/89279737/69f720e6-fb5a-4576-a873-c1bf0d54fe2f)

after (grace):
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/89279737/dbb4f830-2999-44a3-9de8-5871ae373f61)
after (main):
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/89279737/9c91d5b7-2845-485b-a2b0-ae9a14ce3719)

Side-by-side example of text overflow bug and around max font size compensation fix:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/89279737/36fe4c64-48f2-4dc8-95a1-31ac645269ff)
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/89279737/2dca1daf-061b-42e6-96c8-c28b2c717238)

#### known issues
- I couldn't get it to work with singleplayer AI player names. It takes the lobby name instead (BarbarianAI for example). I deprioritized that issue.
